### ISSUE TITLE
WarnOnGenericInferenceFailure reports the same inference failure twice

### ISSUE THREAD

**Issue by vlsi** [uber/NullAway#1735]
### Version

NullAway 0.14.0, and current `master` (21f02dd). Error Prone 2.50.0, JDK 21.

Flags: `-XepOpt:NullAway:JSpecifyMode=true -XepOpt:NullAway:WarnOnGenericInferenceFailure=true -XDaddTypeAnnotationsToSymbol=true`.

### Summary

The same inference failure is reported twice, at the same position of the same file, for every failing call except the first one in a method body.

This is the half of #1725 that #1734 does not address. #1725 shows two reports, one of them in an unrelated file; that misattribution is #1733, and #1734 fixes it. What is left after that fix is the duplicate itself: both copies now print at the same position, so the user sees the identical warning twice. The duplicate does not depend on the misattribution — it reproduces within a single compilation unit, on `master`, and reproduces the same way with #1734 applied.

### Reproducer

```java
// Caller.java
@org.jspecify.annotations.NullMarked
class Caller {
  <T> T call(java.util.function.Supplier<T> s) { return s.get(); }
  void f() {
    call(() -> { return ""; });
    call(() -> { return null; });
  }
}
```

```
/Caller.java:6: warning: [NullAway] inference failure: type variable T constrained to be both @NonNull and @Nullable
    call(() -> { return null; });
        ^
    (see http://t.uber.com/nullaway )
/Caller.java:6: warning: [NullAway] inference failure: type variable T constrained to be both @NonNull and @Nullable
    call(() -> { return null; });
        ^
    (see http://t.uber.com/nullaway )
/Caller.java:6: warning: [NullAway] returning @Nullable expression from method with @NonNull return type
```

### It depends on the position of the call in the method body

Same class, varying the body of `f()`:

| body of `f()` | reports for the failing call |
| --- | --- |
| the failing call alone | 1 |
| a passing call, then the failing call | 2 |
| the failing call, then a passing call | 1 |
| two failing calls | 1 for the first, 2 for the second |

So the first call in the body is reported once and every later failing call twice, whatever the calls around it do.

### Likely cause

`GenericsChecks` reports the failure on both inference passes and only skips caching on one of them. In the `UnsatisfiableConstraintsException` handler, `state.reportMatch(...)` runs whenever `warnOnGenericInferenceFailure()` is set, while the caching below it is guarded by `if (!calledFromDataflow)`: https://github.com/uber/NullAway/blob/21f02ddd3ac3868d5f241ced3eaf4adf8456412c/nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java#L1275-L1294

A call whose inference the dataflow pass reaches before the regular matcher pass does is therefore inferred twice and reported twice. That fits the table above if the regular pass reaches the first call of the body, and caches its result, before dataflow runs over the method.

### Expected

One report per failing call.

### Reproducing it as a test

`CompilationTestHelper` cannot see this: a `// BUG: Diagnostic contains:` line is satisfied by the first matching diagnostic and never fails on a second one. Dropping the comment and reading the "All errors" dump the framework prints is what makes the duplicate visible.
